source/material_model/steinberger.cc

```diff
@@ -578,71 +454,86 @@ namespace aspect
     Steinberger<dim>::evaluate(const MaterialModel::MaterialModelInputs<dim> &in,
                                MaterialModel::MaterialModelOutputs<dim> &out) const
     {
+      double average_temperature(0.0);
+      double average_density(0.0);
+      std::array<std::pair<double, unsigned int>,2> dH;
+
+      if (latent_heat)
+        {
+          for (unsigned int i = 0; i < in.n_evaluation_points(); ++i)
+            {
+              average_temperature += in.temperature[i];
+              average_density += out.densities[i];
```

Comment:
this density is not yet computed, this is why this computation has to happen after the next loop over all points and why the original function had this weird layout with 3 consecutive loops over all points (1. compute point values 2. compute average values 3. compute cell-wise derivative of H). 1 and 2 can probably be combined into one loop, but the cell-wise derivative has to happen afterwards in an extra loop (unless I am missing something).

Reply:
I've now shifted all of the latent_heat calculations into a single if clause at the end of evaluate. This means that density has already been computed, and the function is also easier to read.